cli/flox-rust-sdk/src/providers/publish.rs

```diff
@@ -0,0 +1,148 @@
+use std::path::PathBuf;
+
+use thiserror::Error;
+
+use super::build::ManifestBuilder;
+use super::catalog::Client;
+use super::git::GitCommandProvider;
+use crate::flox::FloxhubToken;
+use crate::models::environment::managed_environment::ManagedEnvironment;
+use crate::models::environment::path_environment::PathEnvironment;
+use crate::models::lockfile::Lockfile;
+
+pub enum PublishEnvironment {
+    Path(PathEnvironment),
+    Managed(ManagedEnvironment),
+}
+
+/// The `Publish` trait describes the high level behavior of publishing a package to a catalog.
+/// Authentication, upload, builds etc, are implementation details of the specific provider.
+/// Modeling the behavior as a trait allows us to swap out the provider, e.g. a mock for testing.
+pub trait Publish {
+    fn publish(
+        &self,
+        client: Client,
+        builder: &dyn ManifestBuilder,
+        environment: PublishEnvironment,
+        package: String,
+    ) -> Result<(), String>;
+}
+
+/// Ensures that the required metadata for publishing is consistent
+#[allow(clippy::manual_non_exhaustive)]
+pub struct CheckedEnvironmentMetadata {
+    pub git_metadata: GitCommandProvider,
+    pub environment_root_relative: PathBuf, // RelativePath
+    pub lockfile: Lockfile,
+
+    // This field isn't "pub", so no one outside this module can construct this struct. That helps
+    // ensure that we can only make this struct as a result of doing the "right thing."
+    _private: (),
+}
+
+impl From<&PathEnvironment> for CheckedEnvironmentMetadata {
```

Comment:
Suggestion: I don't think this is an idiomatic usage of this trait. This trait is most often used to convert between types e.g. `&str -> String` via `String::from("foo")`, whereas this looks like it's extracting information from an environment. A `check_environment_metadata(env:  &PathEnvironment)` function has the same signature as this method, but the semantics are different.

Reply:
I didn't see that this was suggested by @ysndr, but my point still stands that this isn't an idiomatic usage of the trait.

Reply:
Understood.  Do you have a preference between:
- local functions that take a the `enum PublishEnvironment` and return the data 
- OR -
- defining a new trait like `PublishableEnvironment` (or something like that) that is implemented on each Environment type?

I feel like the former keeps Publish stuff where publish is defined and is _simpler_(?), but maybe the later is more Rusty and allows for other Rust centric patterns elsewhere?

Reply:
Chatted with @zmitchell and going with the former, at least until there is a hard reason to do otherwise.